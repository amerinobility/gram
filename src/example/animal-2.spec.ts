import {
  GraphQLBoolean,
  GraphQLInt,
  GraphQLString,
  printSchema,
} from 'graphql'
import { createSchemaBuilder, Service } from '..'

enum AnimalTypes {
  Cat,
  Dog,
}

interface Animal {
  type: AnimalTypes
  name: string
  tame: boolean
  age: number
}

interface Cat extends Animal {
  type: AnimalTypes.Cat
}
interface Dog extends Animal {
  type: AnimalTypes.Dog
}

const Animals: Service<Animal> = {
  findOne: async () => null,
}

const Cats: Service<Cat> = {
  findOne: async ({ order, where }) =>
    Animals.findOne({
      order,
      where: { ...where, type: AnimalTypes.Cat },
    }) as Promise<Cat>,
}
const Dogs: Service<Dog> = {
  findOne: async ({ order, where }) =>
    Animals.findOne({
      order,
      where: { ...where, type: AnimalTypes.Dog },
    }) as Promise<Dog>,
}

describe('testing the example 2', () => {
  it('should build the example code', () => {
    const builder = createSchemaBuilder()
    const animal = builder.interface('Animal')
    // animal name like 'Fluffy', 'Rex'
    // field is not required
    animal.attr('name', GraphQLString)
    // is it a tame animal
    animal.attr('tame', GraphQLBoolean)
    // age of the animal
    animal.attr('age', GraphQLInt)

    const cat = builder.model('Cat', Animals)
    cat.interface('Animal')

    expect(printSchema(builder.build())).toMatchSnapshot()
  })

  it('should build the second part of the example', () => {
    const builder = createSchemaBuilder()
    const animal = builder.interface('Animal', Animals)
    animal.attr('name', GraphQLString)
    animal.attr('tame', GraphQLBoolean)
    animal.attr('age', GraphQLInt)

    const cat = builder.model('Cat', Cats)
    cat.interface('Animal')
    const dog = builder.model('Dog', Dogs)
    dog.interface('Animal')

    expect(printSchema(builder.build())).toMatchSnapshot()
  })
})
