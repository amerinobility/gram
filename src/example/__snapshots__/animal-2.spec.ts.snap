// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing the example 2 should build the example code 1`] = `
"interface Animal {
  name: String
  tame: Boolean
  age: Int
}

type Cat implements Node & Animal {
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
  name: String
  tame: Boolean
  age: Int
}

input CatFilter {
  AND: [CatFilter!]
  OR: [CatFilter!]
  NOT: CatFilter
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  tame: Boolean
  tame_not: Boolean
  age: Int
  age_not: Int
  age_lt: Int
  age_gt: Int
}

enum CatSortOrder {
  id_ASC
  id_DESC
  createdAt_ASC
  createdAt_DESC
  updatedAt_ASC
  updatedAt_DESC
  deletedAt_ASC
  deletedAt_DESC
  name_ASC
  name_DESC
  tame_ASC
  tame_DESC
  age_ASC
  age_DESC
}

input CatWhere {
  AND: [CatFilter!]
  OR: [CatFilter!]
  NOT: CatFilter
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  tame: Boolean
  tame_not: Boolean
  age: Int
  age_not: Int
  age_lt: Int
  age_gt: Int
}

\\"\\"\\"Use JavaScript Date object for date/time fields.\\"\\"\\"
scalar DateTime

interface Node {
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

type Query {
  getCat(where: CatWhere!, order: CatSortOrder): Cat
}
"
`;

exports[`testing the example 2 should build the second part of the example 1`] = `
"interface Animal {
  name: String
  tame: Boolean
  age: Int
}

input AnimalFilter {
  AND: [AnimalFilter!]
  OR: [AnimalFilter!]
  NOT: AnimalFilter
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  tame: Boolean
  tame_not: Boolean
  age: Int
  age_not: Int
  age_lt: Int
  age_gt: Int
}

enum AnimalSortOrder {
  name_ASC
  name_DESC
  tame_ASC
  tame_DESC
  age_ASC
  age_DESC
}

input AnimalWhere {
  AND: [AnimalFilter!]
  OR: [AnimalFilter!]
  NOT: AnimalFilter
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  tame: Boolean
  tame_not: Boolean
  age: Int
  age_not: Int
  age_lt: Int
  age_gt: Int
}

type Cat implements Node & Animal {
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
  name: String
  tame: Boolean
  age: Int
}

input CatFilter {
  AND: [CatFilter!]
  OR: [CatFilter!]
  NOT: CatFilter
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  tame: Boolean
  tame_not: Boolean
  age: Int
  age_not: Int
  age_lt: Int
  age_gt: Int
}

enum CatSortOrder {
  id_ASC
  id_DESC
  createdAt_ASC
  createdAt_DESC
  updatedAt_ASC
  updatedAt_DESC
  deletedAt_ASC
  deletedAt_DESC
  name_ASC
  name_DESC
  tame_ASC
  tame_DESC
  age_ASC
  age_DESC
}

input CatWhere {
  AND: [CatFilter!]
  OR: [CatFilter!]
  NOT: CatFilter
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  tame: Boolean
  tame_not: Boolean
  age: Int
  age_not: Int
  age_lt: Int
  age_gt: Int
}

\\"\\"\\"Use JavaScript Date object for date/time fields.\\"\\"\\"
scalar DateTime

type Dog implements Node & Animal {
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
  name: String
  tame: Boolean
  age: Int
}

input DogFilter {
  AND: [DogFilter!]
  OR: [DogFilter!]
  NOT: DogFilter
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  tame: Boolean
  tame_not: Boolean
  age: Int
  age_not: Int
  age_lt: Int
  age_gt: Int
}

enum DogSortOrder {
  id_ASC
  id_DESC
  createdAt_ASC
  createdAt_DESC
  updatedAt_ASC
  updatedAt_DESC
  deletedAt_ASC
  deletedAt_DESC
  name_ASC
  name_DESC
  tame_ASC
  tame_DESC
  age_ASC
  age_DESC
}

input DogWhere {
  AND: [DogFilter!]
  OR: [DogFilter!]
  NOT: DogFilter
  id: ID
  id_not: ID
  id_in: [ID!]
  id_not_in: [ID!]
  name: String
  name_not: String
  name_in: [String!]
  name_not_in: [String!]
  tame: Boolean
  tame_not: Boolean
  age: Int
  age_not: Int
  age_lt: Int
  age_gt: Int
}

interface Node {
  id: ID
  createdAt: DateTime
  updatedAt: DateTime
  deletedAt: DateTime
}

type Query {
  getAnimal(where: AnimalWhere!, order: AnimalSortOrder): Animal
  getCat(where: CatWhere!, order: CatSortOrder): Cat
  getDog(where: DogWhere!, order: DogSortOrder): Dog
}
"
`;
